rfcs/0000-nixos-containers.md

```diff
@@ -0,0 +1,463 @@
+```
+
+The default page of `nginx` is now reachable like this:
+
+```
+$ curl imperative:80 -i
+$ curl <IPv4 of the host-side veth interface>:8080 -i
+```
+
+#### Modify a container's config imperatively
+
+When `imperative-container.nix` is updated, it can be rebuilt like this:
+
+```
+$ nixos-nspawn update imperative --config ./imperative-container.nix
+```
+
+By default, it will be **restarted**. This can be overridden via `config.activation.strategy`,
+however only `reload`, `restart` and `none` are supported.
+
+Additionally, the way how the container's new config will be activated can be specified
+via `--reload` or `--restart` passed to `nixos-nspawn update`.
+
+If declarative containers are attempted to be modified, the script will terminate early with an
+error.
+
+#### Manage an imperative container's lifecycle
+
+Reboot/Login/etc can be managed via [`machinectl(1)`](https://www.freedesktop.org/software/systemd/man/machinectl.html):
+
+```
+$ machinectl reboot imperative
+$ machinectl shell imperative
+[root@imperative:~]$ …
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Explicit dependency on `networkd` (`networking.useNetworkd = true;`) on both the
+  host-side and container-side.
+  * Since there's a movement to make `systemd-networkd` the default on NixOS, this
+    is from the author's PoV not a big problem.
+
+* Need to migrate from existing containers.
+  * As demonstrated in [*Migration plan*](#migration-plan), a sane path exists.
+  * With a long deprecation time, a rush to migrate can be avoided.
+  * This also means that [the container backend](https://github.com/PsyanticY/nixops-container)
+    for `nixops` needs to be deprecated.
+
+# Alternatives
+[alternatives]: #alternatives
+
+
+* Implement this feature in e.g. its own (optionally community-maintained) repository:
+  * This is problematic due to the changes for [Config activation](#config-activation) that
+    required changes in `switch-to-configuration.pl`.
+* Keep both the proposed feature and the existing `nixos-container` subsystem in NixOS. In contrast
+  to `systemd-nspawn@`, the current container subsystem uses `/var/lib/containers` as state-directory,
+  so clashes shouldn't happen:
+  * The main concern is increased maintenance workload. Also, with the rather prominent
+    name `nixos-container` we shouldn't advertise the old, problematic implementation.
+* Do nothing.
+  * As shown above, this change leverages the full featureset of `systemd-nspawn` and also
+    solves a few existing problems, that are non-trivial to solve when keeping the old
+    implementation.
+  * Since it's planned to move to `networkd` in the longterm anyways, fundamental changes
+    in the container subsystem will be mandatory anyways.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* None that I'm aware of.
+
+# Future work
+[future]: #future-work
+
+* Write documentation for the new module.
+* Get the PR into a mergeable state.
```

Comment:
What PR?

Reply:
https://github.com/NixOS/nixpkgs/pull/140669

Sorry, should've linked it here in the first place :)

Reply:
I just noticed the issue lol. In your YAML header, you refer to the issues like `#number`, but that's parsed as a comment, so it's stripped out in the rendered version... You probably want to use direct links anyway instead of just the issue number, since it's another repository.

Reply:
Done :)